### ISSUE TITLE
Promise intersection with primitive results in wrong awaited type

### ISSUE THREAD

**Issue by Azarattum** [microsoft/TypeScript#48927]
# Bug Report
When we make an intersection of a promise with a primitive type like `Promise<T> & T` (where T is primitive (non-object)), the resulting type from `await (Promise<T> & T)` is wrong (still contains a promise). This is clearly a bug since `Awaited<Promise<T> & T>` and `(Promise<T> & T).then(x => {})` behave correctly (the result is only `T`). The issue is only observed when using the `await` keyword.

The types arrangement might seem arbitrary, but it has a **valid use case** which is provided in the playground link below.

### 🔎 Search Terms

`PROMISE & PRIMITIVE`, `PROMISE INTERSECTION`, `AWAITED TYPE`, `INCONSITENT THEN AND AWAIT`

### 🕗 Version & Regression Information

> Tested only versions available on playground!
- This changed between versions `4.5.5` and `4.7.0-dev.20220503`. `4.4.4` and lower are unaffected by this bug.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=38&ssc=27&pln=38&pc=51#code/PTAEF5K6a0CEBOB7A1gUwHagKIA8BDAWwAcAbdCWa6AWACgCBnAT0wGNQAKASggD5QAbwagxoAC4sSlAIIRQABRREAlk3QAeTAFciAI3SJBAMlC6DRgNyjx7ZJiYTQBBUIC+LpqFk36tsXtHZ30FAgB3AlVnAj9xSWlKeAUpGWQAM1B9K1AQBFAmAAtkHTIAEyzKACILQ0QqgBosnWdo0HVQKuVkNQ1tPTrTcwGjKoBCBgCXADoJQqwuLnY+cEERenixVMoAYRTEjNB2HLy9jpqR+u5wwtV2Qvbve0RELAkeKfceP3dJ+jyaIDwKAAOrIRCoVSYADmuEIpAoVCBsAYzDYnF4AmEU22PgU3V6WgAcpchiTLIg4nYHE4XG5PMwfH4pkFaaFgREojEqVtEvlgdtDtlcmBkudyXUqtdbvdHkdwa9MO8-vECLN5phFsssetNgkZKA9gKDpljiLDXKqhLRtK7g8Os9FcqNuIvj8-gDkbBQeDUAQUDpMBUvSj6LiACrIADy+gAVuh2BJNOHBMDw6B0HgJFgyt4nIgodDbAB+UAAZQkBZhtgAXKB05ns0HvLUjCXQNbELX6xmsznvPpkMgKARMO34EOR2OXXWG33mwUWAZh+2y0vB2Ru+HmWG+TsevooegysmmgBVBQp-EqdRaK9mSMx+OJzRn-g7tEcbgrNY4vnyYF9wMI8T1bYweXlYI6WBDwvCZFVqSg9kXEiNpYj-A1kmNNJMmFPIxW8C4KSlLgbjtOVHTeD5-BdMQ1TmBYlh-bFaPEXEjX1dBDjNU5LTAkiyNlB0FSoz5vgYX4aM9EM4AAGVUNQJAICRVAcJEZPAVFWC-TFVhY+JWWcVgiAUNdlzILgqkDVQAEcdCqcTWNxADDQPED+gpJowNAAAfTjDmM98WRpGJ6Tg3wELEVRMi4VxIEXIg+F1PVzRBSgyGQWlRxcQ8yGiFhJGQLwNEQZwGNAay7MoYyNwKKF2EocJKCCAA3IxswqNoJCKszauQOMEwkT5IsgtkwlQ7kMKSfYcKyE5RT4y4BJle0nhEpVqKmeiNS1ZjkvidiZq4015otc5+NtIS1peUTWLdCSPTADS4DPDRDWYShnsgLT0W-HVgqgsplLi4R5jIDK6yqcJwXKKp3AgwzQAyggykLUz12HSzkdRmEHIg3EABFgaOw4gaU-G+Vk5AUbR7DjqR6mcehHcDv-a8elvTQiaUoYuG51w-KpmmYUcgyQug4QGW8CKaPiaLuFcMZgWxwskqmVVpjBjLzX5uVsePdXxG2hY8B1L5zXgFpQCagoVPB0BCgINrJHmUASBvN78x0RMdFeQqkfUJtQHScEKpIMn0CYYbZcQsaOQmlwKcwkncNO5InFUe2giUqFCKFpmpW2Bopn0TX0HB5BTqYIqbf9ShAwIfREW6iq3qqMuK-b0AAHJ4G78sAAkozPWSCYQHBcAARTPWRZPrKMe52fuowAJR7gnu4mVj8IKYpSgqIko3TbOomwKp88LQu+SYerKGUl2OhIZAoVaVpvGhHR-VHbNj39wwGZRseLeW11SMW1HpfabE9wpyOKdM43hYw6FpFUbm6BlrkWEjdDaht2Ad21nkNKRxspIOqHg5AVQxIs3EHkIkyBsxNDmPfWQE1jzJkEIYR2bVrqvETAHDAi9u7F23mAAoRV2AjkQGQAqvd+4dH0CgDAmBphTVAGPYEzCuSsNkIICSQA)

### 💻 Code

```ts
type A = Promise<number> & number;
const a = {} as A;

const b = await a;
type B = typeof b; // B should be "number", but it is "Promise<number> & number"!

a.then((c) => {
  type C = typeof c; // C is "number" (which is corrent)
});

type D = Awaited<A> // D is "number" (which is corrent)
```

### 🙁 Actual behavior
Type `B` still contains `Promise<number>` despite of the value being awaited.

### 🙂 Expected behavior
In the example above `B` **has to be equal** to `C` and `D`.

**Comment by Azarattum** [comment]
https://github.com/microsoft/TypeScript/pull/45350 this is probably related. However, as shown before, the `Awaited` itself works correctly.

**Comment by fatcerberus** [comment]
I don’t understand (even with your playground example) why you would ever write `Promise<number> & number`.  This type represents a value that is simultaneously both a promise and a primitive number, which is impossible.

**Comment by jcalz** [comment]
The code in there seems to be using `symbol`, but there are no values which are both symbols and promise objects.  Could you clearly articulate the use case in words?  

Even better, could you show a working example of your use case at runtime? Your playground link uses `{}` as the runtime value which won't actually work at runtime.  Like, forget the TS types for a bit, what kind of JS behavior are you trying to support?

**Comment by andrewbranch** [comment]
It does seem theoretically wrong, but yeah, unless there is a real-world use case I doubt we’ll prioritize a change here.

**Comment by Andarist** [comment]
I don't think this would be a hard fix - at the moment this is somewhat explicitly called out to be invalid in one of the comments:
https://github.dev/microsoft/TypeScript/blob/8f56f6b49d2fb282ff95df64f62dab2947004e95/src/compiler/checker.ts#L36317-L36321

**Probably** one of the required changes to make this work would be to tweak the related logic in the `getPromisedTypeOfPromise` here:
https://github.dev/microsoft/TypeScript/blob/8f56f6b49d2fb282ff95df64f62dab2947004e95/src/compiler/checker.ts#L36265-L36268


I could take a stab at this - if only somebody would confirm that this behavior needs to be adjusted. I also don't see though how this could be utilized in the real-world code so far.

**Comment by andrewbranch** [comment]
I did a quick experiment monkeypatching `String.prototype.then` to see if it would be called by `Promise.resolve` (or `await`, which uses the algorithm of the former). It did not. It’s pretty hard to follow but I think I found this in the spec: https://tc39.es/ecma262/#sec-promise-resolve-functions

> 8. If [Type](https://tc39.es/ecma262/#sec-ecmascript-data-types-and-values)(resolution) is not Object, then
>   a. Perform [FulfillPromise](https://tc39.es/ecma262/#sec-fulfillpromise)(promise, resolution).
>   b. Return undefined.

The only way I can interpret `string & Promise<string>` is a value that is definitely the primitive `string` but also manages to satisfy the `Promise` interface structurally, which perhaps you could do with a horrifying regimen of monkeypatching. The spec and my experiment seem to agree that such a value (the stuff of nightmares) would in fact emerge unchanged from `await`. So the type is correct.

**Comment by Azarattum** [comment]
Thank you all for your feedback.

@andrewbranch I don't think this issue should be closed. I just made a [full example use case](https://www.typescriptlang.org/play?target=7#code/PQKhAIHkCMCsFMDGAXcB3ATgQwA4-huFgM7oD2GA1lhmQK4B2AJuCMALABQyAnvuABUyMBCgA8AgHzgAvIPDwAHsnjNSxZBgCWDAOZdw4APzgAypp37OhgFzylKteAZ0AttAIHj4AHJuPGF52AgrKqkyk0Fq6OsheJgBC0QCSDHHW4MGhjhHg0GRkADbwWAzx4AkFxaVBggDcXFygEABKWogAFuA4tK5axPDgTFjIWOC8-Ggd7V2IZO468KQ9ZPgYyFpL4GQAZkQMPMjTeuhaR0Tdvf2DbFwTg22dAApXAxIANOAAqtJyX9nhUiMJjwHaLJjlF7za4SSS1f4OQHOOiFQqQ17wWG1KF9N5ScAAMkEwjgSGQYh+DU4TTA4CehSwiHgHSKIMIxB47iK4B2FHQHVU4wFGEG-WcZCIKDoWEKQxGYx0F20nUu0IGrA4nDmDA04EKZCwTEssjMnPyhQAFAAiAAyBqNeitAEoqTSIABhVabUhjFbIMj3bZ7UrbUkoAB0BgghnA7pFIy2vto-sDnSwirQZy6+WBPMYKC0ZB1PL5PS0fQ2ADclpHrNHDABBPCFb3bHAbIsy7bVjDaEGkf1CwZpnS11gx8AteDIOgYYshshh1BHEbgRAhjzgOgDFgkC5+gN8eBjsCanb5jsMVX+xD6hjwC3EegYJl2Uo8T5kHt9pZ2KdzDAmDEd9PnfSQnXAABvLxtV1A8TREMlw10acoRTI9IB2C1EJQR9nyZJ0XRgotdVGDAUNQOQcOQcNEHjFQLQPIiMmo8MBmQNDD3wTCLTIijPiY10Ml5QgLVg1AVhwIMoCXZDp0gNAGChNZeB8LBXCWRjkzIQioK8QwtD2Xij12cAn1nJlwAAQjkK1F1EZBnT0jIJ3EqDwErGU6HgT5+LM6dwAAX1kfSJxkhy5OQBSlNoFSeAAESWOitHbCgtIDMgBNiiCAB8cqgwKqTC8BDPAYz8FMzzCm82RbPPBgCyLJzoJcsLWJBMF72UgheF4mg-Mkz5II8rz4DsKrvPDKJmDwiz4AgwLmOKwLQoMoyKPAPL-OQCCWuKwx2tBRZuvWHg+vI6cstWIbVrCii7Aoowpp0JhZpfeb3luid2Lsdinum17zPep1PtaidFqKsKVta6HDGh4idVQW8i3gFg5GG8NMa-Agf1IQqvFY9jOPuHjkfvJhPj46clvAEUZznNc71Rql4c4Zoo1jeZpq2SSetbUyxhiasr2GUZTnOJM1WPDmp3p+dnHgNBaZmVVcWluszwvQsr1UZUOg+b5JHStW7BxGEpE+UWsDsL5doR3VdZmVGENkuiSgYlxURp0ryvgUyrdq2z7LJJzWJIYhogYC1Hc6VHLflZj7dQNAaCvOQLQgmRpD2mNYKKY8U7nC1boAAwEDAeGNQdGSZYgfT1e1jWDlArJL8AAGovvANuAFksB4TceHoEtyIDC4AHIAB0ABIA5shvDUsce1wFRBKCMEvQppww6dnK8rQAbQ5LlZTtRe9AAXStSH8epDI3Ox3stH7E0c8MA-TDNIpw39F5yzOLQ1YL52AzrIaQVtQYTgmvATCIDM7SAtAHPc74nRPWgZhJ6oCtoQNCv6cw2g9BwLAWVRB8oiA+gOJnOQ+pz66G8IXKOEE7D4MsKQ0YhFIExiOKoOwKw1Y-wFAwZ6M0+HXBBqFdcyBOi8IxLREYnRhGvVEQMcRrVOoykKDwGRUtwzqNRDwRRxsxGcNvkna8Eo5AHjJg+CB3YcbPyWDTQmqFtIkywjHAUFNzE013gzDxzs9yPA6GbPEnxKRcFZuzOs4Ae5kDXhcdcqJxgSjGCKVwAZBgNieMkKMmpihIw0aQOQAAGKkJAeANTzA1S8RAcBaHdBo0BOdEmFGIO3dudRwDAGAGYcsKIEy03gOklQW4cCi3gF4LAKczgKyViEh8hAs7bQEOWP2dBkAWgwJ8AATMUvZhFIY+xaUUuQABGCCvj97EDoGsEqrgcAUFGGkOUoxr5eHgK0wYlzwBWifBpFepQUIsCtm8zgkSekAHF9TQC7BoAZaY9DOwtNuZ2wpBjV0rGQZ+a4SL5zMjgdSTouD5IBYitGPIZQDCElE8cDYG6rHFl0bZQxDI7AIKoVAIpmAEDMpoBMMQay5K4OUyp9VGpXi5WyO0qxHz+mwChP8Mx5liA0AQ3QnxVXGnyvcUyNCHS6EkBqvlKhdBaPAKcza4Btl2wyD7KyCKgUXOnHvSGDrnZyB2JS+AkNmhmAEC0BsAgACiEKACa5qbD6XrOAAAtOKF5YwPAiUGHqywUaJxxoDmKG54yWBTEFPeJWWafSeS0AyaAxQo2ai8D7VV-KeCB3NTaic3TwAAHVBgp2eYOTcWAK3oolCisyX9ZRlgrIArYiptSwC1kWRlqtrjEE4YYVtZkJRoGHCGWO8SRJDgXvq8YR4xwxm6aFVtCQkBYCHRu8AHQsDVjsU-JgIIrwfxHT-Mgf9x1AKFGKbdlAz09KGEdBgZwPkNpUKiIEpBSo3rvQ++96Zy3FATdsdkI7ugMiZCyQobJQq1rlVgFC1lqGNz0M2sKraGyoHfOMVZ-JBTFv3c7G9uhpTYDSJsFgGgKBEYHXkUUiNSiWVMiwvQt087FHDOHHqsrePEZ0HChqfs9hid0NvE9PTTDrs3VeFp4ANJHDILkUsBARKuDbAQEY2tiASdxVJ-Uug5PyuPMQHALYNlWh+SDH58V5R2C8+3Xl8n5qhWCh89Ub9c72ePI560Z99WY3DM6MLXhWZrTKnWk1DaZC5atRRldtJTD+sDSG8N2zI1g2jXGhgEoA5JooCmsjVgwrVdQ7gfANAfTqAKFePcerUarTa-mq8haF3qhoNoasS6FBpB5amk4cKVC+QlIOHo8BKyFm3Ly+AOAhuahbT05IrgNJGgTJo2bKhCDcP3caJbgxeICgM+mEWrL2XKZ5L0PdWX4CmqbaFElxAWRoAS8aOQmhvKQ0MOxFZGl6AbNAUsqLGXHzA9B+RnFOp87hji7aZrSXnRQ6CkS1qrapxPkKAh5g21ZkJtCjxlzAjVBsLGEj26QOyAg+ayaT1nyieGEk7FsgTmrZsXc2ca03nPhWj86MALHcE0aYho0UnPTcvq415rrX2ude671-ruQggAASQaKhfAhYYA3MhAOGBvZS9dpRlwSl7f25J6BtAjJu3MEERB8gPpbJQMaZ6QASJizj4X8X8eY0J6HxGqG5BTPTKgBnfGieC-DyL+UYuPOS+dNL2X1sfkK6thp24qvDDQHWegQYNBBiMD7ShwcTAJRPlAoUIzdBdBdBmWKRz7QNENocZxxJNurdj-HxPnXhgBAm7NxbwOeu0sRJVzSiAQbFDqXczX5s-eal98QEKzgIrEBVPFWZYH64BiyuwNluwFr8rWuciurTNyHnqjg-emvwWXMMZFNX9Ajubugeu2Xgq6fygwAcOwX2qSQyGStSWg4Y4ANo04486g-ce6WSyQgycs9cmqegx6gOhGCqk4SqGIKqFgegRqaqlqOqewC2BqQkz+SBKBaBDaH+CGgy3KhA+oDK4B+af+K4ycjuA4ZAoBQGHOSsN2kBX2z43+fGx6bkPBUkcg7EqQV2VUFox+ZU8CT+MYieMykqBA0qOAzmfGVB9aNMgUnwpyeyxSicGQrabagi-+YIGAuoeqqGUB8w+w8B8ajIM4XYOg5m1mc6-QYhhgiGZaDegm32RBPka6-+OAVe489By87mjIzIrIPKx85oQ4IoZiyigwCedSDSqIGckMKexGcgHiRiAwnw9BNMbq5KEO3q4RDYOwV2e6hRt6e4IoFO1YeagwSRqAsh-h0osoWaV4N2es42Pk4R-IZagwwBdSJwN2FA0QOgXY54hQYIqIzshRiB2me6Ac64V4vaiAtcO4eQAYHQ8xXhFmMx3RpQLAax2gMQDAXY3Rj48AnaHQbBgw94zsghJUUxT2hh3BBQOAJOAuJEqAAcCe0yEkGIFRcRJoNRhR8c7Crqd6ZKJoLRjBXSPSyByAqBZk6BWAHRPKYwUwixaGW4AwGApJ-h2s-+XaTuoy4yxx8o8xRwYomYSSEhxxb2XKlkHgyAG6go0x4Q2Rxqv23okyiJtO8ymyxCMOqy8OmyOythByZiha8yJouA9SjSNMlRRRs2esFoepGImJWAjROJQKeJGAkOKuMJcehaBeBpipVpUsKJIWaJDAFp3pasnw7pCc2JgK7q4wTprRd8ERXpisdIGIKpSyapcO6ympVq2pjR1QGAahBAGhShzErMmhHIlSiO2c9s2OuOpgF+JATWtCvKN+v2DapyMeGQ+hyetZV+5yrqhSJopSlZDmEeVoNZnOl+9ZB6P2f22ybZcZSe5+Y5dZFo1qLMToGcXAQAA) which roughly replicates what I'm trying to achieve in my project. I'm working on a custom asynchronous [store for Svelte](https://svelte.dev/tutorial/custom-stores) similar to [this one](https://github.com/sveltetools/svelte-asyncable) or [this one](https://github.com/square/svelte-store). The idea behind it is to abstract away API interactions, so the store can manage everything. To achieve that I'm using a data structure which I call `RichPromise` which is basically `Promise<T> & T` of anything (including primitives). This is achieved by constructing a special prototype chain which includes the original object's prototype, bound methods for primitives, overrides such things as `[Symbol.toPrimitive]` and `valueOf`, and adds `then`, `catch`, `finally` to mimic a promise behavior (for implementation details see `enrich` function in the example). Because the original prototype is included in RichPromise's chain `enrich(Promise.resolve("2"), "1") instanceof String` is true.

This has one caveat though. When we first look at out rich promise (before any api calls resolve), there is no data to be concatenated with a promise, so we pair it with `loading` symbol placeholder. This way the type becomes `Promise<T> & (T | typeof loading)`. I choose to use a symbol, so we can infer the loading state by type with `if (richPromise == loading)` which works because of the `[Symbol.toPrimitive]` override. So, it is must have for primitive types to work correctly because the guard above only works if we use `unique symbol`, but not `Symbol` which makes sense. However as we do not have `unique Symbol` as an object there is no other way to solve it.

This whole approach was chosen to elegantly support 2 different rendering strategies in svelte templates:
```svelte
<!-- Strategy 1 -->
<p>
  {#if $store != loading}
    {$store}
  {:else}
    <!-- Is shown only the first time -->
    Loading
  {/if}
</p>

<!-- Strategy 2 -->
<p>
  {#await $store}
    <!-- Is shown every time new data starts loading -->
    Loading
  {:then data}
    {data}
  {/await}
</p>
```

In the linked example I'm trying to mimic something similar to svelte stores without the framework. If you run the code, you'll see the showcases of the two strategies shown above:
```
Showcase loading strategy 1
Loading...
(3) ['super', 'important', 'data'] 'Data: super important data'
(3) ['super', 'important', 'data'] 'Data: super important data'
(3) ['some', 'changed', 'data'] 'Data: some changed data'

Showcase loading strategy 2
Loading...
(3) ['super', 'important', 'data'] 'Data: super important data'
(3) ['super', 'important', 'data'] 'Data: super important data'
Loading...
(3) ['some', 'changed', 'data'] 'Data: some changed data'
(3) ['some', 'changed', 'data'] 'Data: some changed data'
```
Note that in a real scenario the same data would not be rendered twice.

Anyway besides all of that, this is INCONSISTENCY (between `then`, `await`, `Awaited`) and **should be considered a bug**. Moreover, since this was introduced just recently it cannot be seen an intended behavior.

**Comment by Azarattum** [comment]
Also, if we had a way to
```ts
data: A | B;
if (data != B) {
  data: A;
}
```
where `B extends object`. That would also solve my particular use case. But I would rather get the original issue solved for the sake of consistency.

If anyone knows if the snippet above is achievable (without changing an operator or using a function) please let me know.

<details>
  <summary>Fun fact</summary>

  I also tried to use `document.all` and `%GetUndetectable()` for prototyping a falsy object to a promise, so this would be possible:

  ```svelte
{#if $store}
    {$store}
{:else}
    Loading
{/if}

{#await $store}
    Loading
{:then data}
    {data}
{/await}
  ```
 But it was a little bit to far... And there were some problems with promises not being detected since `typeof document.all` returns `undefined` and not an `object`. So I decided to go ahead with the symbol approach.
</details>

**Comment by Azarattum** [comment]
And even ignoring everything considered, it just makes sense for `await Promise<number> & number` to return `number`. Regardless of if anybody is going to use it or not...

**Comment by Azarattum** [comment]
@andrewbranch Yeah... I think you are right after all. I did use an `Object(<value>)` wrapper when constructing my promises. So the types are actually a `String` and not a `string`. And I even think I'm fine with that even how it works now.

But what I actually want is to have `unique Symbol` or rather  `unique object` in general. So we can narrow based on equality to singleton objects.
```typescript
const singleton = {} as unique Record<string, any>;

const unknown = {a: "test"} as {a: string} | typeof singleton.

if (unknown != singleton) {
  unknown: {a: "test"};
  unknown.a; //test
}
```
If there's already a way to do this in the language, please let me know. If not, I'll open a feature request then.

**Comment by fatcerberus** [comment]
`unique object` sounds like nominal typing: see e.g. https://github.com/microsoft/TypeScript/issues/202
